### ISSUE TITLE
S3 Sink fails with error - Unsupported record :java.math.BigDecimal

### ISSUE THREAD

**Issue by parepallykiran** [Landoop/stream-reactor#994]
# Issue Guidelines

Please review these questions before submitting any issue?

### What version of the Stream Reactor are you reporting this issue for?

Using commit sha `dd85a361469c4d9a563ab5e32c47e37d75c8df51`

### Are you running the correct version of Kafka/Confluent for the Stream reactor release?

We are running on AWS MSK Connect. v2.7.1

### What is the expected behaviour?

As per our configuration, there supposed to be no headers in the message except String header added by TimeStampConverter SMT. Messages should be written to S3, without any issues. 

### What was observed?

But Sink is failing with Bigdecimal conversion errors. 

### What is your Connect cluster configuration (connect-avro-distributed.properties)?

Using default AWS MSK Connect configs with 1 task and 1 MCU

### What is your connector properties configuration (my-connector.properties)?

```
connector.class=io.lenses.streamreactor.connect.aws.s3.sink.S3SinkConnector
transforms.formattedEventTs.format.to.pattern=yyyy-MM-dd-HH
transforms.formattedEventTs.unix.precision=nanoseconds
tasks.max=1
topics=topic-1
connect.s3.error.policy=NOOP
transforms=formattedEventTs
transforms.formattedEventTs.header.name=eventTs
key.converter.schemas.enable=false
connect.s3.kcql=insert into s3-test-bucket:metrics select * from `topic-1` PARTITIONBY _value.`customer_id`, _header.eventTs  STOREAS `JSON` WITHPARTITIONER=Values WITH_FLUSH_SIZE=1000000 WITH_FLUSH_INTERVAL=30 WITH_FLUSH_COUNT=5000
name=lensesS3Sink
value.converter.schemas.enable=false
transforms.formattedEventTs.field=_value.timestamp
connect.s3.aws.region=us-east-1
transforms.formattedEventTs.target.type=string
value.converter=org.apache.kafka.connect.json.JsonConverter
errors.log.enable=true
key.converter=org.apache.kafka.connect.storage.StringConverter
transforms.formattedEventTs.type=io.lenses.connect.smt.header.TimestampConverter
```

### Please provide full log files (redact and sensitive information)

```
[Worker-02213660f692aeb50] [2023-10-18 23:36:12,606]  Encountered error [Unsupported record 4.7755507E+3751563:java.math.BigDecimal] (io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask:61)
[Worker-02213660f692aeb50] org.apache.kafka.connect.errors.ConnectException: Unsupported record 4.7755507E+3751563:java.math.BigDecimal
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.conversion.ValueToSinkDataConverter$.apply(ValueToSinkDataConverter.scala:47)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.conversion.HeaderToStringConverter$.$anonfun$apply$1(HeaderToStringConverter.scala:26)
[Worker-02213660f692aeb50] 	at scala.collection.StrictOptimizedIterableOps.map(StrictOptimizedIterableOps.scala:100)
[Worker-02213660f692aeb50] 	at scala.collection.StrictOptimizedIterableOps.map$(StrictOptimizedIterableOps.scala:87)
[Worker-02213660f692aeb50] 	at scala.collection.convert.JavaCollectionWrappers$JIterableWrapper.map(JavaCollectionWrappers.scala:62)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.conversion.HeaderToStringConverter$.apply(HeaderToStringConverter.scala:25)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask.$anonfun$put$4(S3SinkTask.scala:156)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask.$anonfun$put$4$adapted(S3SinkTask.scala:145)
[Worker-02213660f692aeb50] 	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:575)
[Worker-02213660f692aeb50] 	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:573)
[Worker-02213660f692aeb50] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:933)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask.$anonfun$put$1(S3SinkTask.scala:145)
[Worker-02213660f692aeb50] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[Worker-02213660f692aeb50] 	at scala.util.Try$.apply(Try.scala:210)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask.put(S3SinkTask.scala:134)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:586)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:329)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:232)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:189)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:238)
[Worker-02213660f692aeb50] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[Worker-02213660f692aeb50] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[Worker-02213660f692aeb50] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[Worker-02213660f692aeb50] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[Worker-02213660f692aeb50] 	at java.base/java.lang.Thread.run(Thread.java:829)
[Worker-02213660f692aeb50] [2023-10-18 23:36:12,606] WARN  Error policy NOOP: [Unsupported record 4.7755507E+3751563:java.math.BigDecimal]. Processing continuing. (com.datamountaineer.streamreactor.common.errors.NoopErrorPolicy:56)
[Worker-02213660f692aeb50] [2023-10-18 23:36:12,628] ERROR  Encountered error [Unsupported record 8.79565E+586389113:java.math.BigDecimal] (io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask:61)
[Worker-02213660f692aeb50] org.apache.kafka.connect.errors.ConnectException: Unsupported record 8.79565E+586389113:java.math.BigDecimal
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.conversion.ValueToSinkDataConverter$.apply(ValueToSinkDataConverter.scala:47)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.conversion.HeaderToStringConverter$.$anonfun$apply$1(HeaderToStringConverter.scala:26)
[Worker-02213660f692aeb50] 	at scala.collection.StrictOptimizedIterableOps.map(StrictOptimizedIterableOps.scala:100)
[Worker-02213660f692aeb50] 	at scala.collection.StrictOptimizedIterableOps.map$(StrictOptimizedIterableOps.scala:87)
[Worker-02213660f692aeb50] 	at scala.collection.convert.JavaCollectionWrappers$JIterableWrapper.map(JavaCollectionWrappers.scala:62)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.conversion.HeaderToStringConverter$.apply(HeaderToStringConverter.scala:25)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask.$anonfun$put$4(S3SinkTask.scala:156)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask.$anonfun$put$4$adapted(S3SinkTask.scala:145)
[Worker-02213660f692aeb50] 	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:575)
[Worker-02213660f692aeb50] 	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:573)
[Worker-02213660f692aeb50] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:933)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask.$anonfun$put$1(S3SinkTask.scala:145)
[Worker-02213660f692aeb50] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[Worker-02213660f692aeb50] 	at scala.util.Try$.apply(Try.scala:210)
[Worker-02213660f692aeb50] 	at io.lenses.streamreactor.connect.aws.s3.sink.S3SinkTask.put(S3SinkTask.scala:134)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:586)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:329)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:232)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:189)
[Worker-02213660f692aeb50] 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:238)
[Worker-02213660f692aeb50] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[Worker-02213660f692aeb50] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[Worker-02213660f692aeb50] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[Worker-02213660f692aeb50] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[Worker-02213660f692aeb50] 	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by aljoshare** [comment]
Do you need more information @stheppi? Or is it already clear whats causing this? I'm currently debugging and trying to find the value which is causing the error.

**Comment by aljoshare** [comment]
It seems that the error is thrown when trying to process a record with a UUID (e.g. Jaeger trace-id) as a header value. Can you confirm this? @parepallykiran

**Comment by parepallykiran** [comment]
@aljoshare Looked at the headers of my events and they are being added by APM agent with random UUID's and trace ID's. For sure some of these headers are causing issues.

**Comment by stheppi** [comment]
@parepallykiran I am not sure how the UUID are ending up converting BigDecimal. Connect is handing over the data to the sink and one of the headers is received as java BigDecimal. I don't understand how the implementation of org.apache.kafka.connect.storage.HeaderConverter would translate from Bytes to BigDecimal.  Can you please share an example of you kafka message? 

Also, @parepallykiran  you should be using the latest 5.0 release and use the `store.envelope`=true. See the docs [here](https://docs.lenses.io/5.3/connectors/sinks/s3sinkconnector/).

**Comment by parepallykiran** [comment]
> @parepallykiran I am not sure how the UUID are ending up converting BigDecimal. Connect is handing over the data to the sink and one of the headers is received as java BigDecimal. I don't understand how the implementation of org.apache.kafka.connect.storage.HeaderConverter would translate from Bytes to BigDecimal. Can you please share an example of you kafka message?
> 
> Also, @parepallykiran you should be using the latest 5.0 release and use the `store.envelope`=true. See the docs [here](https://docs.lenses.io/5.3/connectors/sinks/s3sinkconnector/).

@stheppi Sorry for confusion here. We see there are couple headers auto added by OpenTelemetry SDK. Some of these headers are UUID's and couple random IDs. Connect is failing to convert some of the headers and failing to write to S3. I'm not sure which exact messages are getting dropped as only few messages are failing to write and don't have a failed message logged. Anything you can recommend to capture failed message processing? 

We are using latest branch from Stream reactor. We don't want to store headers and keys to S3. Do we still need to set store.envelope=true ?

**Comment by stheppi** [comment]
>We are using latest branch from Stream reactor. We don't want to store headers and keys to S3. Do we still need to set store.envelope=true ?
Then no. 

I wonder if opentelemetry stores some of the metrics as bigdecimal and not double. I still don't know how the default HeaderConverter would lift that to bigdecimal. 
Can you please confirm your header.converter value?

**Comment by parepallykiran** [comment]
> > We are using latest branch from Stream reactor. We don't want to store headers and keys to S3. Do we still need to set store.envelope=true ?
> > Then no.
> 
> I wonder if opentelemetry stores some of the metrics as bigdecimal and not double. I still don't know how the default HeaderConverter would lift that to bigdecimal. Can you please confirm your header.converter value?

@stheppi we are not setting any of the header.converter value.

**Comment by parepallykiran** [comment]
> > > We are using latest branch from Stream reactor. We don't want to store headers and keys to S3. Do we still need to set store.envelope=true ?
> > > Then no.
> > 
> > 
> > I wonder if opentelemetry stores some of the metrics as bigdecimal and not double. I still don't know how the default HeaderConverter would lift that to bigdecimal. Can you please confirm your header.converter value?
> 
> @stheppi we are not setting any of the header.converter value.

@stheppi Sample headers in the Kafka message added by our tracing agent.

![Screenshot 2023-11-08 at 10 27 41 AM](https://github.com/lensesio/stream-reactor/assets/5224319/46a41957-c19c-4091-9de1-fa535d45893c)

**Comment by stheppi** [comment]
Strange. A PR is in place so soon you can try it.

**Comment by teddymaef** [comment]
We got a similar error for java.util.Date. Seems testing not complete for all the common types? can someone check & add tests and support for all common types?

org.apache.kafka.connect.errors.ConnectException: Unsupported record Wed Aug 02 12:45:58 GMT 2023:java.util.Date

**Comment by stheppi** [comment]
This has been fixed and latest version 6.3.0 covers it.
